### DISCUSSION TITLE
Properly read server-side environment variables

### DISCUSSION THREAD

**Discussion by salieflewis** [vercel/next.js#58388] [App Router]
I have secrets that I don't want exposed to the browser, and are only required to be used in the context of my Server Actions. From my understanding, these environment variables need not be prefixed by NEXT_PUBLIC, and I can simply include them in my `.env.local` file. However, I'm running into issues where certain functions of the code are still being run in the browser because they are not asynchronous and can't be coupled with my Server Actions. 

Is this where the `server-only` package comes into play?

The workaround that I've found is exposing the environment variable in my `next.config.js` like so:

```
module.exports = {
  env: {
    PRIVATE_KEY: process.env.PRIVATE_KEY,
  },
}
```

However this worries me because I don't have a thorough understanding of what is going on given I couldn't find this anywhere in the docs. Any insight would be greatly appreciated, thank you.

**Comment by jhackGC** [comment]
That will expose the vars to both client and server side code. Dont do it if you dont want that.

It happened to me a few times and it was because I had to be 100% sure that the retrieval of the non NEXT_PUBLIC env var was being done in a server context, either a server action or an api route, it won't be retrieved if its running in a client side context. 

You should have it printed in the browser console as a server log, hope it helps.
and , BTW, server actions dont work in a pages architecture, only work with apps architecture, dont mix them up, pages with routes, apps with either routes or server actions.
Cheers
